sql/src/main/java/org/apache/druid/sql/DirectStatement.java

```diff
@@ -62,8 +65,81 @@ public class DirectStatement extends AbstractStatement implements Cancelable
 {
   private static final Logger log = new Logger(DirectStatement.class);
 
+  /**
+   * Represents the execution plan for a query with the ability to run
+   * that plan (once).
+   */
+  public class ResultSet implements Cancelable
```

Comment:
is a name like `PlannerResultWrapper` or something similar more suitable? because this class seems to augment the functionalities of `PlannerResult` with some statement related accessors.

Reply:
The name echos the JDBC concept of a result set. It is a way to get results for a statement. The idea, in JDBC, is that rather than the statement having two "modes" (running or not), the lifecycle is encoded in the class structure. A statement does the work of everything up to a set of results; the result set then manages those results. As in JDBC, some operations are common (close, etc.)

This is needed because of a use case (outside of Druid itself) where we want to manage the results in a different thread than the one that prepares the statement. Splitting out the functionality to a `ResultSet` is a bit cleaner than having "is prepared and ready to run" state in the statement itself.

This is a quick & dirty refactoring to solve issues created by the previous PR. As you point out, it could be improved after we give it a bit more thought.

Reply:
Thanks a lot for explanation! Yes, I agree that the semblence to JDBC concept makes sense and I was also thinking in the same line. But I was thinking that maybe then the `run` or `execute` method should be a part of the statement instead of the resultset. Maybe, the result-set could start with a `Yielder` object over the result sequence.

But agree that all that could be done after more thought as a follow up. And completely agree with the concept of breaking Statement into Statement and ResultRet :+1: